sql: TestReadCommittedImplicitPartitionUpsert failed

sql.TestReadCommittedImplicitPartitionUpsert [failed](https://mesolite.cluster.engflow.com/invocations/default/22970c6b-d179-4098-9f35-e06b4ce15b1b?testReportRun=6&testReportShard=7&testReportAttempt=1#targets-Ly9wa2cvc3FsOnNxbF90ZXN0) on master @ [8bb7919bc6bf36fa175cd7d30629e6d884d41630](https://github.com/cockroachdb/cockroach/commits/8bb7919bc6bf36fa175cd7d30629e6d884d41630):


```
=== RUN   TestReadCommittedImplicitPartitionUpsert
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestReadCommittedImplicitPartitionUpsert826828863
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:141: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
Starting test 1 -- "UPSERT INTO d.upsert VALUES (1, 10, 'east', 100)"
Starting test 2 -- "INSERT INTO d.upsert VALUES (1, 10, 'west', 101) ON CONFLICT DO NOTHING"
    mutation_test.go:254: pq: duplicate key value violates unique constraint "upsert_pkey"
    panic.go:635: -- test log scope end --
test logs left over in: outputs.zip/logTestReadCommittedImplicitPartitionUpsert826828863
--- FAIL: TestReadCommittedImplicitPartitionUpsert (0.78s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=6</code>
 - <code>shard=7</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReadCommittedImplicitPartitionUpsert.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql.TestReadCommittedImplicitPartitionUpsert [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20484537?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20484537?buildTab=artifacts#/) on master @ [942c0cf1e2cfa7718caf6c2f7399a2a8624b63eb](https://github.com/cockroachdb/cockroach/commits/942c0cf1e2cfa7718caf6c2f7399a2a8624b63eb):


```
=== RUN   TestReadCommittedImplicitPartitionUpsert
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/004ac2cdb3e4dcce971cd33db6280729/logTestReadCommittedImplicitPartitionUpsert625365964
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:141: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
Starting test 1 -- "UPSERT INTO d.upsert VALUES (1, 10, 'east', 100)"
Starting test 2 -- "INSERT INTO d.upsert VALUES (1, 10, 'west', 101) ON CONFLICT DO NOTHING"
    mutation_test.go:254: pq: duplicate key value violates unique constraint "upsert_pkey"
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/004ac2cdb3e4dcce971cd33db6280729/logTestReadCommittedImplicitPartitionUpsert625365964
--- FAIL: TestReadCommittedImplicitPartitionUpsert (1.21s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReadCommittedImplicitPartitionUpsert.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/exec_util.go
+++ b/pkg/sql/exec_util.go
@@ -2017,7 +2017,7 @@ type ExecutorTestingKnobs struct {
 
 	// AfterArbiterRead, if set, will be called after each row read from an arbiter index
 	// for an UPSERT or INSERT.
-	AfterArbiterRead func()
+	AfterArbiterRead func(query string)
 
 	// BeforeIndexSplitAndScatter is invoked with the split and scatter of an index
 	// occurs.

--- a/pkg/sql/insert.go
+++ b/pkg/sql/insert.go
@@ -326,7 +326,7 @@ func (n *insertNode) BatchedNext(params runParams) (bool, error) {
 		if buildutil.CrdbTestBuild {
 			// This testing knob allows us to suspend execution to force a race condition.
 			if fn := params.ExecCfg().TestingKnobs.AfterArbiterRead; fn != nil {
-				fn()
+				fn(params.p.stmt.SQL)
 			}
 		}
 

--- a/pkg/sql/insert_fast_path.go
+++ b/pkg/sql/insert_fast_path.go
@@ -476,7 +476,7 @@ func (n *insertFastPathNode) BatchedNext(params runParams) (bool, error) {
 		if buildutil.CrdbTestBuild {
 			// This testing knob allows us to suspend execution to force a race condition.
 			if fn := params.ExecCfg().TestingKnobs.AfterArbiterRead; fn != nil {
-				fn()
+				fn(params.p.stmt.SQL)
 			}
 		}
 

--- a/pkg/sql/mutation_test.go
+++ b/pkg/sql/mutation_test.go
@@ -10,6 +10,7 @@ import (
 	gosql "database/sql"
 	"fmt"
 	"reflect"
+	"strings"
 	"sync"
 	"testing"
 
@@ -133,13 +134,20 @@ func TestReadCommittedImplicitPartitionUpsert(t *testing.T) {
 		mu.l.Unlock()
 	}
 
+	peekState := func() State {
+		mu.l.Lock()
+		defer mu.l.Unlock()
+		return mu.state
+	}
+
 	ctx := context.Background()
 	params, _ := createTestServerParamsAllowTenants()
 	// If test is in Ready state, transition to ReadDone and wait for conflict.
 	params.Knobs = base.TestingKnobs{
 		SQLExecutor: &sql.ExecutorTestingKnobs{
-			AfterArbiterRead: func() {
-				if mu.state != Ready {
+			AfterArbiterRead: func(query string) {
+				// Only wait for arbiter operations on the upsert table.
+				if peekState() != Ready || !strings.Contains(query, "d.upsert") {
 					return
 				}
 				setState(ReadDone)

--- a/pkg/sql/upsert.go
+++ b/pkg/sql/upsert.go
@@ -207,7 +207,7 @@ func (r *upsertRun) processSourceRow(params runParams, rowVals tree.Datums) erro
 	if buildutil.CrdbTestBuild {
 		// This testing knob allows us to suspend execution to force a race condition.
 		if fn := params.ExecCfg().TestingKnobs.AfterArbiterRead; fn != nil {
-			fn()
+			fn(params.p.stmt.SQL)
 		}
 	}
 

